### PR TITLE
Retrofit TG pathfinder with upstream improvements

### PR DESCRIPTION
**Pull request by warriorstar-orion**
## What Does This PR Do
This PR updates our implementation of SSpathfinder with upstream improvements, particularly:

- https://github.com/tgstation/tgstation/pull/70623
- https://github.com/tgstation/tgstation/pull/72685
- https://github.com/tgstation/tgstation/pull/78684

Read those PRs for more details, but the high-level overview is:

- SSpathfinder is now a firing subsystem which works to spread pathfinding operations across ticks
- `/datum/can_pass_info` is introduced. It's essentially a small struct containing all the information needed to determine the passability of an atom without having to pass the atom itself around. It's used to simplify the `CanAStarPass` (`CanPathfindPass` in our code) API, reduce or eliminate the need to look up the caller itself while performing the checks, and allows for creating cacheable, reusable map passability data.

There is some use of the `src.` syntax in this code, I may clean this up but there are places where it does improve readability (proc args having the same names as their assigned fields).

Don't ask me about the Binding of Isaac variable names in `/proc/can_pass_check_vars`, I don't even know.

## Why It's Good For The Game
There are many performance gains suggested by the above PRs. It's also foundational work for porting TG AI controllers.

## Testing
- [x] Tested secbot patrol
- [x] Tested secbot arrest
- [x] Tested medibot patrol
- [x] Tested medibot help
- [x] Tested Mulebot delivery
- [x] Tested parrot pathfinding to objects of interest
- [x] Tested changeling stings (yes they pathfind)
- [x] Test pathfinder modsuit
- [x] Test maint drone recall pathfinding

Note that hostile mobs have their internal implementation of "pathfinding" that won't use this until AI controllers are ported.

## Changelog
NPFC

### CODE REVIEW

**Comment by 1080pCat** [comment]
> * Tested changeling stings (yes they pathfind)

????????

hopefully this improves haunted objects just not working.

File: code/__DEFINES/turfs.dm [hunk 1]
```diff
@@ -4,3 +4,9 @@
 #define TURF_PATHING_PASS_PROC 1
 /// Turf is never passable
 #define TURF_PATHING_PASS_NO 2
+
+/// Returns a list of turfs similar to CORNER_BLOCK but with offsets
+#define CORNER_BLOCK_OFFSET(corner, width, height, offset_x, offset_y) ((block(locate(corner.x + offset_x, corner.y + offset_y, corner.z), locate(min(corner.x + (width - 1) + offset_x, world.maxx), min(corner.y + (height - 1) + offset_y, world.maxy), corner.z))))
```
Comment (Contrabang) [inline code/__DEFINES/turfs.dm:9]: block now accepts direct coords, and its faster
image
https://www.byond.com/docs/ref/#/proc/block

**Comment by S34NW** [comment]
> > * Tested changeling stings (yes they pathfind)
> 
> ????????

You can't just do a range check because it won't account for glass panes etc. Easiest way is pathfind

File: code/__DEFINES/turfs.dm (hunk 1 above)
Comment (warriorstar-orion) [reply code/__DEFINES/turfs.dm]: Fixed, PTAL

**Comment by SteelSlayer** [comment]
Seeing various "Invalid JPS destination" stack traces pop up:

A majority seem to be cleanbot related.
```
Runtime in code/__HELPERS/stacktrace.dm,14: Invalid JPS destination
[2024-06-19T13:19:31]   proc name: stack trace (/datum/proc/stack_trace)
[2024-06-19T13:19:31]   src: /datum/pathfind/jps (/datum/pathfind/jps)
[2024-06-19T13:19:31]   call stack:
[2024-06-19T13:19:31]   /datum/pathfind/jps (/datum/pathfind/jps): stack_trace
[2024-06-19T13:19:31]   /datum/pathfind/jps (/datum/pathfind/jps): start
[2024-06-19T13:19:31]   Pathfinder (/datum/controller/subsystem/pathfinder): pathfind
[2024-06-19T13:19:31]   get path to(Cleanbot (/mob/living/simple_animal/bot/cleanbot), the blood (/obj/effect/decal/cleanable/blood/footprints), 30, null, /list (/list), 1, null, 1, 2)
[2024-06-19T13:19:31]   Cleanbot (/mob/living/simple_animal/bot/cleanbot): handle_automated_action
[2024-06-19T13:19:31]   NPC Pool (/datum/controller/subsystem/npcpool): fire
[2024-06-19T13:19:31]   NPC Pool (/datum/controller/subsystem/npcpool): ignite
[2024-06-19T13:19:31]   Master (/datum/controller/master): RunQueue
[2024-06-19T13:19:31]   Master (/datum/controller/master): Loop
[2024-06-19T13:19:31]   Master (/datum/controller/master): StartProcessing
```
```
Runtime in code/__HELPERS/stacktrace.dm,14: Invalid JPS destination
[2024-06-20T00:22:10]   proc name: stack trace (/datum/proc/stack_trace)
[2024-06-20T00:22:10]   usr: Alicia Carroll (CKEY) (/mob/living/carbon/human)
[2024-06-20T00:22:10]   usr.loc: The floor (146,102,2) (/turf/simulated/floor/plasteel)
[2024-06-20T00:22:10]   src: /datum/pathfind/jps (/datum/pathfind/jps)
[2024-06-20T00:22:10]   call stack:
[2024-06-20T00:22:10]   /datum/pathfind/jps (/datum/pathfind/jps): stack_trace
[2024-06-20T00:22:10]   /datum/pathfind/jps (/datum/pathfind/jps): start
[2024-06-20T00:22:10]   Pathfinder (/datum/controller/subsystem/pathfinder): pathfind
[2024-06-20T00:22:10]   get path to(Tayachi (/mob/living/carbon/human), null, 2, null, /list (/list), 0, null, 1, 2)
[2024-06-20T00:22:10]   spread germs by incision(the left arm (/obj/item/organ/external/arm), the cautery (/obj/item/cautery/augment))
[2024-06-20T00:22:10]   spread germs to organ(the left arm (/obj/item/organ/external/arm), Alicia Carroll (/mob/living/carbon/human), the cautery (/obj/item/cautery/augment))
[2024-06-20T00:22:10]   cauterize incision (/datum/surgery_step/generic/cauterize): begin_step
[2024-06-20T00:22:10]   cauterize incision (/datum/surgery_step/generic/cauterize): begin_step
[2024-06-20T00:22:10]   cauterize incision (/datum/surgery_step/generic/cauterize): initiate
[2024-06-20T00:22:10]   cauterize incision (/datum/surgery_step/generic/cauterize): try_op
[2024-06-20T00:22:10]   Manipulate Organs (proxy) (/datum/surgery_step/proxy/manipulate_organs): try_next_step
[2024-06-20T00:22:10]   Manipulate Organs (proxy) (/datum/surgery_step/proxy/manipulate_organs): try_op
[2024-06-20T00:22:10]   Organ Manipulation (/datum/surgery/organ_manipulation_boneless): next_step
[2024-06-20T00:22:10]   Tayachi (/mob/living/carbon/human): attackby
[2024-06-20T00:22:10]   Tayachi (/mob/living/carbon/human): attackby
[2024-06-20T00:22:10]   the cautery (/obj/item/cautery/augment): melee_attack_chain
[2024-06-20T00:22:10]   Alicia Carroll (/mob/living/carbon/human): ClickOn
[2024-06-20T00:22:10]   Tayachi (/mob/living/carbon/human): Click
```

**Comment by warriorstar-orion** [comment]
@SteelSlayer after taking a look, I think it's acceptable to simply remove this runtime. This can happen whenever a target that was just found is removed asynchronously, for example a cleanbot pathing to a mess that was just cleaned up. Because this stuff can happen asynchronously a guard in the critical path is only so useful:

https://github.com/ParadiseSS13/Paradise/blob/5f2b7d7cd04430e67283901ad38397148ce5ccf4/code/modules/mob/living/simple_animal/bot/cleanbot.dm#L157-L167

`target` can become null between lines 157 and 160. Since these are just as valid a failure to path as the other conditions in `/datum/pathfind/jps/start()` I think it's fine to just kill the stack trace. Thoughts?

**Comment by SteelSlayer** [comment]
@warriorstar-orion Yeah that seems good with me.

**Pull request merged**